notesystem/modes/search_mode.py

```diff
@@ -0,0 +1,202 @@
+
+
+class SearchMode(BaseMode[SearchModeArguments]):
+    """Search markdown files (notes) for the given search terms"""
+
+    def _run(self, args) -> None:  # TODO: Should return exit code
+        """Entry point for search mode
+
+        Sets the state, and starts the search
+
+        Arguments:
+            args {SearchModeArguments} -- The arguments from the parser
+        Raises:
+            {FileNotFoundError} -- When the given path cannot be found
+        """
+
+        if 'tag_str' in args and args['tag_str'] is not None:
+            self.tags = args['tag_str'].split(' ')
+        else:
+            self.tags = []
+        self.topic = args['topic']
+        self.title = args['title']
+        self.case_insensitive = args['case_insensitive']
+        self.pattern = args['pattern']
+        self.path = args['path']
+        self.matches: List[SearchMatch] = []
+
+        if os.path.isfile(self.path):
+            self._search_file(self.path)
+        elif os.path.isdir(self.path):
+            self._search_dir(self.path)
+        else:
+            raise FileNotFoundError(f'{self.path} could not be found')
+
+        # Print out the results
+        if self._visual:
+            c = 0
+            for match in self.matches:
+                for _ in match['matched_lines']:
+                    c += 1
+
+            print(
+                colored('Found', 'cyan'),
+                colored(str(c), 'cyan', attrs=['bold']),
+                colored('results: ', 'cyan'),
+            )
+            for match in self.matches:
+                print_search_result(match, self.pattern)
+
+    def _parse_frontmatter(self, file_lines: List[str]) -> Dict[str, str]:
+
+        fm_data: Dict[str, str] = {}
+
+        if not file_lines[0].startswith('---'):
+            return fm_data
+
+        for line in file_lines[1:]:
+            if line.startswith('---'):
+                break
+            try:
+                key, value = line.strip().split(':')
+                fm_data[key.strip().lower()] = value.strip()
+            except ValueError:
+                # It is no valid frontmatter
+                break
+
+        return fm_data
+
+    def _search_file(self, file_path: str) -> None:
+        """Search through the given file and adds matches to the matches list
+
+        Search for the pattern in given file (file_path) if tags are given
+        that need to be matched they are checked first to prevent looking
+        throught the whole file.
+
+        If an match is found a SearchMatch (dict) is added to the matches list.
+
+        Arguments:
+            file_path {str} -- The path of the file to search through
+        """
+
+        title = None
+        topic = None
+        tags = None
+
+        with open(file_path, 'r') as file:
+            lines = file.readlines()
+
+        if not len(lines) >= 1:
+            return  # Emtpy file
+
+        # Check if there is front matter
+        if lines[0].startswith('---'):
```

Comment:
Does front matter always start with `---` and is it always at the start of the file?

Reply:
yes